File: src/main/java/epicsquid/superiorshields/network/NetworkHandler.java
package epicsquid.superiorshields.network;

import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.function.Supplier;

import epicsquid.superiorshields.SuperiorShields;
import net.minecraft.network.PacketBuffer;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.network.NetworkEvent;
import net.minecraftforge.fml.network.NetworkRegistry;
import net.minecraftforge.fml.network.simple.SimpleChannel;
import net.minecraftforge.network.NetworkEvent;
import net.minecraftforge.network.NetworkRegistry;
import net.minecraftforge.network.simple.SimpleChannel;

public class NetworkHandler {

	private static final String PROTOCOL_VERSION = "1";
	public static final SimpleChannel INSTANCE = NetworkRegistry.ChannelBuilder
			.named(new ResourceLocation(SuperiorShields.MODID, "main"))
			.clientAcceptedVersions(PROTOCOL_VERSION::equals)
			.serverAcceptedVersions(PROTOCOL_VERSION::equals)
			.networkProtocolVersion(() -> PROTOCOL_VERSION)
			.simpleChannel();

	private static int id = 0;

	public static void register() {
		registerMessage(SPacketShieldUpdate.class, SPacketShieldUpdate::encode, SPacketShieldUpdate::decode, SPacketShieldUpdate::handle);
	}

	private static <M> void registerMessage(Class<M> messageType, BiConsumer<M, PacketBuffer> encoder, Function<PacketBuffer, M> decoder, BiConsumer<M, Supplier<NetworkEvent.Context>> messageConsumer) {
		INSTANCE.registerMessage(id++, messageType, encoder, decoder, messageConsumer);
	}
}

File: src/main/kotlin/dev/epicsquid/superiorshields/network/NetworkHandler.kt
package dev.epicsquid.superiorshields.network

import dev.epicsquid.superiorshields.SuperiorShields
import dev.epicsquid.superiorshields.registry.CapabilityRegistry.shield
import net.minecraft.client.Minecraft
import net.minecraft.resources.ResourceLocation
import net.minecraftforge.network.NetworkRegistry
import net.minecraftforge.network.simple.SimpleChannel

object NetworkHandler {
	const val PROTOCOL_VERSION = "1.0"

	private var id = 0
	private fun nextId(): Int = id++

	val channel: SimpleChannel by lazy {
		NetworkRegistry.newSimpleChannel(
			ResourceLocation(SuperiorShields.MODID, "network"),
			{ PROTOCOL_VERSION },
			{ true },
			{ true })
	}

	// TODO wtf is going on here? Is it actually a false positive?
	@Suppress("INACCESSIBLE_TYPE")
	fun register() {
		channel.registerMessage(
			nextId(),
			SuperiorShieldUpdatePacket::class.java,
			{ msg, byteBuf -> encodeTo(byteBuf, msg) },
			::decodeFrom,
			{ msg, ctx ->
				ctx.get().apply {
					enqueueWork {
						Minecraft.getInstance().player?.let { player ->
							player.shield.hp = msg.hp
						}
					}
					packetHandled = true
				}
			}
		)
	}
}